File: 2003/CS101SamplesAll/CS101SamplesVSTO/CreatingMenusInOutlook/CreatingMenusInOutlook/ThisApplication.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Microsoft.VisualStudio.Tools.Applications.Runtime;
using Outlook = Microsoft.Office.Interop.Outlook;
using Office = Microsoft.Office.Core;

namespace CreatingMenusInOutlook
{
    public partial class ThisApplication
    {
        Office.CommandBar mainMenuBar;
        Office.CommandBarPopup newMenuEntry;
        Office.CommandBarButton inboxCountCommand, addButtonCommand;
        List<Office.CommandBarButton> newButtonCommands =
            new List<Microsoft.Office.Core.CommandBarButton>();
        Office._CommandBarButtonEvents_ClickEventHandler newButtonClickHandler;

        const string MENU_BEFORE = "Help";
        int helpMenuIndex;

        Outlook.MAPIFolder inBoxfolder;

        private void ThisApplication_Startup(object sender, System.EventArgs e)
        {
            // Obtain a reference to the Inbox.  This will be used
            // later for message counts.
            inBoxfolder = this.GetNamespace("MAPI").GetDefaultFolder(Outlook.OlDefaultFolders.olFolderInbox);

            // Obtain a reference to the menu bar.  This will be the
            // container for the new options.
            mainMenuBar = this.ActiveExplorer().CommandBars.ActiveMenuBar;

            // The helpMenuIndex is used to establish where the new
            // menu will appear (in this instance, before the Help menu.
            helpMenuIndex = mainMenuBar.Controls[MENU_BEFORE].Index;

            // Create the new menu.  It will display as "Outlook Custom Menu"
            newMenuEntry = (Office.CommandBarPopup)mainMenuBar.Controls.Add(
                Office.MsoControlType.msoControlPopup,
                missing, missing, helpMenuIndex, true);
            newMenuEntry.Caption = "Outlook Custom Menu";
            newMenuEntry.Visible = true;

            // Add menu items using the Controls.Add() method of the
            // menu.  The Caption properties control the visible label.
            inboxCountCommand = (Office.CommandBarButton)newMenuEntry.Controls.Add(
                Office.MsoControlType.msoControlButton,
                missing, missing, missing, true);
            inboxCountCommand.Caption = "Inbox Count";
            inboxCountCommand.Visible = true;
            inboxCountCommand.Enabled = true;
            
            // Standard event binding is used to perform an action when
            // the menu option is clicked by the user.
            inboxCountCommand.Click +=
                new Office._CommandBarButtonEvents_ClickEventHandler(inboxCountCommand_Click);

            // Add a menu command to dynamically add new buttons
            addButtonCommand = (Office.CommandBarButton)newMenuEntry.Controls.Add(
                Office.MsoControlType.msoControlButton,
                missing, missing, missing, true);
            addButtonCommand.Caption = "Add new CommandBarButton";
            addButtonCommand.Visible = true;
            addButtonCommand.Enabled = true;

            // Standard event binding is used to perform an action when
            // the menu option is clicked by the user.
            addButtonCommand.Click +=
                new Office._CommandBarButtonEvents_ClickEventHandler(addButtonCommand_Click);

            // This event handler will be reused between all instances
            // of the dynamically-added menu commands.
            newButtonClickHandler = 
                new Office._CommandBarButtonEvents_ClickEventHandler(newButtonCommand_Click);
        }

        /// <summary>
        /// Handles the click event for the add button command.
        /// This creates a new command bar button in the same menu.
        /// </summary>
        /// <param name="Ctrl"></param>
        /// <param name="CancelDefault"></param>
        void addButtonCommand_Click(Microsoft.Office.Core.CommandBarButton Ctrl, ref bool CancelDefault)
        {
            Office.CommandBarButton newButtonCommand;

            // Add a menu command to dynamically add new buttons
            newButtonCommand = (Office.CommandBarButton)newMenuEntry.Controls.Add(Office.MsoControlType.msoControlButton, missing, missing, missing, true);
            newButtonCommand.Caption = "Button #" + newMenuEntry.Controls.Count;
            newButtonCommand.Visible = true;
            newButtonCommand.Enabled = true;

            // Retain a reference to the control or event binding will not
            // occur once the object goes out of scope.
            newButtonCommands.Add(newButtonCommand);

            // Standard event binding is used to perform an action when
            // the menu option is clicked by the user.
            newButtonCommand.Click += newButtonClickHandler;
        }

        /// <summary>
        /// This is invoked each time a dynamic command bar button is clicked.
        /// It simply displays a message box with the control's name.
        /// </summary>
        /// <param name="Ctrl"></param>
        /// <param name="CancelDefault"></param>
        void newButtonCommand_Click(Microsoft.Office.Core.CommandBarButton Ctrl, ref bool CancelDefault)
        {
            MessageBox.Show("You clicked " + Ctrl.Caption);
        }

        /// <summary>
        /// The event handler to be invoked when the Inbox Count
        /// menu command is clicked.
        /// </summary>
        /// <param name="Ctrl"></param>
        /// <param name="CancelDefault"></param>
        void inboxCountCommand_Click(Office.CommandBarButton Ctrl, ref bool CancelDefault)
        {
            MessageBox.Show("Inbox currently contains: " + inBoxfolder.Items.Count + " items (" + inBoxfolder.UnReadItemCount + " unread)");
        }

        private void ThisApplication_Shutdown(object sender, System.EventArgs e)
        {
        }

        #region VSTO generated code

        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InternalStartup()
        {
            this.Startup += new System.EventHandler(ThisApplication_Startup);
            this.Shutdown += new System.EventHandler(ThisApplication_Shutdown);
        }

        #endregion
    }
}

File: 2003/VB101SamplesAll/VB101SamplesVSTO/CreatingMenusInOutlook/CreatingMenusInOutlook/ThisApplication.vb
Imports System.Collections.Generic
Imports Outlook = Microsoft.Office.Interop.Outlook
Imports Office = Microsoft.Office.Core

public class ThisApplication

    Dim mainMenuBar As Office.CommandBar
    Dim newMenuEntry As Office.CommandBarPopup
    Dim addButtonCommand, inboxCountCommand As Office.CommandBarButton
    Dim newButtonClickHandler As Office._CommandBarButtonEvents_ClickEventHandler
    Dim newButtonCommands As New List(Of Office.CommandBarButton)

    Dim MENU_BEFORE As String = "Help"
    Dim helpMenuIndex As Integer
    Dim inBoxfolder As Outlook.MAPIFolder

    Private Sub ThisApplication_Startup(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Startup
        ' Obtain a reference to the Inbox.  This will be used
        ' later for message counts.
        inBoxfolder = Me.GetNamespace("MAPI").GetDefaultFolder( _
            Outlook.OlDefaultFolders.olFolderInbox)

        ' Obtain a reference to the menu bar.  This will be the
        ' container for the new options.
        mainMenuBar = Me.ActiveExplorer.CommandBars.ActiveMenuBar

        ' The helpMenuIndex is used to establish where the new
        ' menu will appear (in this instance, before the Help menu.
        helpMenuIndex = mainMenuBar.Controls(MENU_BEFORE).Index

        ' Create the new menu.  It will display as "Outlook Custom Menu"
        newMenuEntry = CType(mainMenuBar.Controls.Add( _
            Office.MsoControlType.msoControlPopup, Before:=helpMenuIndex, Temporary:=True), _
            Office.CommandBarPopup)
        newMenuEntry.Caption = "Outlook Custom Menu"
        newMenuEntry.Visible = True

        ' Add menu items using the Controls.Add() method of the
        ' menu.  The Caption properties control the visible label.
        inboxCountCommand = CType(newMenuEntry.Controls.Add( _
            Office.MsoControlType.msoControlButton, Temporary:=True), _
            Office.CommandBarButton)
        inboxCountCommand.Caption = "Inbox Count"
        inboxCountCommand.Visible = True
        inboxCountCommand.Enabled = True

        ' Standard event binding is used to perform an action when
        ' the menu option is clicked by the user.
        AddHandler inboxCountCommand.Click, AddressOf Me.inboxCountCommand_Click

        ' Add a menu command to dynamically add new buttons
        addButtonCommand = CType(newMenuEntry.Controls.Add( _
            Office.MsoControlType.msoControlButton, Temporary:=True), _
            Office.CommandBarButton)
        addButtonCommand.Caption = "Add new CommandBarButton"
        addButtonCommand.Visible = True
        addButtonCommand.Enabled = True

        ' Standard event binding is used to perform an action when
        ' the menu option is clicked by the user.
        AddHandler addButtonCommand.Click, AddressOf Me.addButtonCommand_Click

        ' This event handler will be reused between all instances
        ' of the dynamically-added menu commands.
        newButtonClickHandler = AddressOf Me.newButtonCommand_Click
    End Sub

    ''' <summary>
    ''' Handles the click event for the add button command.
    ''' This creates a new command bar button in the same menu.
    ''' </summary>
    ''' <param name="Ctrl"></param>
    ''' <param name="CancelDefault"></param>
    Private Sub addButtonCommand_Click(ByVal Ctrl As Microsoft.Office.Core.CommandBarButton, ByRef CancelDefault As Boolean)
        Dim newButtonCommand As Office.CommandBarButton

        ' Add a menu command to dynamically add new buttons
        newButtonCommand = CType(newMenuEntry.Controls.Add( _
            Office.MsoControlType.msoControlButton, Temporary:=True), _
            Office.CommandBarButton)
        newButtonCommand.Caption = "Button #" & newMenuEntry.Controls.Count
        newButtonCommand.Visible = True
        newButtonCommand.Enabled = True

        ' Retain a reference to the control or event binding will not
        ' occur once the object goes out of scope.
        newButtonCommands.Add(newButtonCommand)

        ' Standard event binding is used to perform an action when
        ' the menu option is clicked by the user.
        AddHandler newButtonCommand.Click, newButtonClickHandler
    End Sub

    ''' <summary>
    ''' This is invoked each time a dynamic command bar button is clicked.
    ''' It simply displays a message box with the control's name.
    ''' </summary>
    ''' <param name="Ctrl"></param>
    ''' <param name="CancelDefault"></param>
    Private Sub newButtonCommand_Click(ByVal Ctrl As Microsoft.Office.Core.CommandBarButton, ByRef CancelDefault As Boolean)
        MessageBox.Show("You clicked " & Ctrl.Caption)
    End Sub

    ''' <summary>
    ''' The event handler to be invoked when the Inbox Count
    ''' menu command is clicked.
    ''' </summary>
    ''' <param name="Ctrl"></param>
    ''' <param name="CancelDefault"></param>
    Private Sub inboxCountCommand_Click(ByVal Ctrl As Office.CommandBarButton, ByRef CancelDefault As Boolean)
        MessageBox.Show("Inbox currently contains: " _
                        & inBoxfolder.Items.Count & " items (" _
                        & inBoxfolder.UnReadItemCount & " unread)")
    End Sub

    Private Sub ThisApplication_Shutdown(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Shutdown
    End Sub

End Class
